target/linux/ath79/dts/ar7161_meraki_mr16.dts

```diff
@@ -0,0 +1,189 @@
+#include <dt-bindings/input/input.h>
+
+#include "ar7100.dtsi"
+
+/ {
+	compatible = "meraki,mr16", "qca,ar7161";
+	model = "Meraki MR16";
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	aliases {
+		led-boot = &power_orange;
+		led-failsafe = &power_orange;
+		led-running = &power_green;
+		led-upgrade = &power_orange;
+	};
+
+	extosc: ref {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-output-names = "ref";
+		clock-frequency = <40000000>;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		wifi1 {
+			label = "mr16:green:wifi1";
+			gpios = <&gpio 0 GPIO_ACTIVE_LOW>;
+		};
+
+		wifi2 {
+			label = "mr16:green:wifi2";
+			gpios = <&gpio 1 GPIO_ACTIVE_LOW>;
+		};
+
+		wifi3 {
+			label = "mr16:green:wifi3";
+			gpios = <&gpio 2 GPIO_ACTIVE_LOW>;
+		};
+
+		wifi4 {
+			label = "mr16:green:wifi4";
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
+		};
+
+		wan {
+			label = "mr16:green:wan";
+			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+		};
+
+		power_orange: power_orange {
+			label = "mr16:orange:power";
+			gpios = <&gpio 5 GPIO_ACTIVE_LOW>;
+			panic-indicator;
+		};
+
+		power_green: power_green {
+			label = "mr16:green:power";
+			gpios = <&gpio 6 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 7 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&pcie0 {
+	status = "okay";
+
+	ath9k0: wifi@0,11 { /* 2.4 GHz */
+		compatible = "pci168c,0029";
+		reg = <0x8800 0 0 0 0>;
+		qca,no-eeprom;
+		mtd-mac-address = <&mac 0x66>;
+		mtd-mac-address-increment = <1>;
+		#gpio-cells = <2>;
+		gpio-controller;
+	};
+
+	ath9k1: wifi@0,12 { /* 5 GHz */
+		compatible = "pci168c,0029";
+		reg = <0x9000 0 0 0 0>;
+		qca,no-eeprom;
+		mtd-mac-address = <&mac 0x66>;
+		mtd-mac-address-increment = <2>;
+		#gpio-cells = <2>;
+		gpio-controller;
+	};
+};
+
+&pll {
+	clocks = <&extosc>;
+};
+
+&uart {
+	status = "okay";
+};
+
+&mdio0 {
+	status = "okay";
+
+	phy-mask = <0x1>;
+
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+	};
+};
+
+&eth0 {
+	status = "okay";
+	mtd-mac-address = <&mac 0x66>;
+
+	pll-data = <0x00110000 0x00001099 0x00991099>;
+
+	phy-mode = "rgmii";
+	phy-handle = <&phy0>;
+};
+
+&spi {
+	status = "okay";
+	num-cs = <1>;
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x40000>;
+				read-only;
+			};
+
+			partition@40000 {
+				label = "u-boot-env";
+				reg = <0x40000 0x40000>;
+				read-only;
+			};
+
+			mac: partition@80000 {
+				label = "mac";
+				reg = <0x80000 0x10000>;
+				read-only;
+			};
+
+			partition@90000 {
+				label = "firmware";
+				reg = <0x90000 0xf50000>;
+				compatible = "fixed-partitions";
+
+				partition@0 {
+					label = "rootfs";
+					reg = <0x000000 0xd20000>;
+				};
+
+				partition@d20000 {
```

Comment:
So, this will effectively move the kernel partition back by 64k compared to ar71xx. If that doesn't matter, i.e. the kernel partition's position is not required to be fixed, can't we just use some splitter here and go back to dynamic partitioning?

Reply:
Could you point me to some documentation or some other keyword I could use to peruse DTS files to see syntax?

Reply:
This is specifically in context of the word "splitter".

Reply:
I haven't reviewed a case so far where kernel wasn't first in the firmware partition. If kernel is first, its start offset of course is fixed by the firmware partitions start, and then positioning the rootfs will just be matter of kernel size. This rootfs positioning can then be done by one the splitters here: https://github.com/openwrt/openwrt/tree/master/target/linux/generic/files/drivers/mtd/mtdsplit

Selection is done by adding the corresponding compatible to the firmware partition. This essentially is a mechanism to remove the necessity to resize kernel partitions manually from time to time when the kernel has grown beyond partition limits.

In your case, _I don't think_ this will make sense, as if the kernel partition is second, its offset cannot be moved around as for the rootfs in my earlier example; and I also don't know whether the splitters will work in that case (swapped order for kernel/rootfs).
For now, it might just work to implement fixed partitions with the correct offset.
(This will work until the kernel outgrows its partition size of about 2.2 MiB (I don't know how big the kernel is on ath79/generic ATM, but for ath79/nand several devices went beyond the 2 MiB limit). After that limit is reached, you will either have to implement a multi-stage bootloader that will be able to start the actual kernel from another position, or have the device set to DEFAULT := n and will have to build with local config manually again to reduce kernel size. Long story, but the essence is: If you are already close to that limit, it might be better to implement a decent solution right away.)

Reply:
All makes sense. I removed the forced kernel-last config order from the makefile  and used compatible='denx,uboot'. Works fine.

Reply:
Interesting. So, where does the bootloader learn the kernel offset from then?

Reply:
We have to set it in u-boot manually; from the commit: `setenv bootcmd bootm 0xbf0a0000; saveenv;`

Reply:
This should be resolved.